File: packages/xyd-opencli2rust/src/flags.ts
// clap Arg emission for flags + positionals (the flags.ts mirror from opencli2go).
// Bool flags accept both `--x` and `--x=false` (num_args 0..=1 + require_equals),
// matching urfave/cli's bool semantics; slices accept repetition and commas.

import type { FlagModel } from './model';
import { chain, lit, rsStr, type RsVal } from './rslit';

/** A clap `.num_args(...)` range from an OpenCLI arity, or null when unconstrained. */
function arityRange(arity: { minimum?: number; maximum?: number }): string | null {
  const hasMin = typeof arity.minimum === 'number';
  const hasMax = typeof arity.maximum === 'number';
  if (hasMin && hasMax) return `${arity.minimum}..=${arity.maximum}`;
  if (hasMin) return `${arity.minimum}..`;
  if (hasMax) return `..=${arity.maximum}`;
  return null;
}

/** A clap `.value_parser([...])` argument literal from a list of accepted values. */
function valueParserList(values: string[]): RsVal {
  return lit(`[${values.map((v) => JSON.stringify(v)).join(', ')}]`);
}

export function renderFlagArg(f: FlagModel): RsVal {
  const calls: [string, RsVal[]][] = [['long', [rsStr(f.flagName)]]];

  const shorts = f.aliases.filter((a) => a.length === 1);
  const longs = f.aliases.filter((a) => a.length > 1);
  if (shorts.length) calls.push(['short', [lit(`'${shorts[0]}'`)]]);
  if (longs.length === 1) calls.push(['visible_alias', [rsStr(longs[0])]]);
  else if (longs.length > 1) calls.push(['visible_aliases', [lit(`[${longs.map((l) => JSON.stringify(l)).join(', ')}]`)]]);

  switch (f.flagType) {
    case 'int':
      calls.push(['value_parser', [lit('clap::value_parser!(i64)')]]);
      break;
    case 'float':
      calls.push(['value_parser', [lit('clap::value_parser!(f64)')]]);
      break;
    case 'bool':
      calls.push(['num_args', [lit('0..=1')]]);
      calls.push(['require_equals', [lit('true')]]);
      calls.push(['default_missing_value', [rsStr('true')]]);
      calls.push(['value_parser', [lit('clap::value_parser!(bool)')]]);
      break;
    case 'slice':
      calls.push(['action', [lit('clap::ArgAction::Append')]]);
      calls.push(['value_delimiter', [lit("','")]]);
      break;
    default:
      break; // string | json | file: clap's default String parsing
  }

  // Non-API leaf extras — additive and field-gated. `buildLeafModel` never sets
  // these, so x-openapi flags render exactly as before.
  if (f.acceptedValues?.length) calls.push(['value_parser', [valueParserList(f.acceptedValues)]]);
  if (f.arity) {
    const range = arityRange(f.arity);
    if (range) calls.push(['num_args', [lit(range)]]);
  }

  if (f.description) calls.push(['help', [rsStr(f.description)]]);
  if (f.required) calls.push(['required', [lit('true')]]);
  if (f.hidden) calls.push(['hide', [lit('true')]]);

  return chain(`Arg::new(${JSON.stringify(f.flagName)})`, calls);
}

export function renderFlagArgs(flags: FlagModel[]): RsVal[] {
  return flags.map(renderFlagArg);
}

/**
 * A positional argument (clap requires declaring positionals; Go read them untyped).
 * `local` gates the non-API extras (`acceptedValues → .value_parser`, `arity →
 * .num_args`): the x-openapi call site passes `false`, so API positionals stay
 * byte-identical even when their OpenCLI argument carries those fields.
 */
export function renderPositionalArg(
  arg: { name: string; required?: boolean; description?: string; acceptedValues?: string[]; arity?: { minimum?: number; maximum?: number } },
  local: boolean,
): RsVal {
  const calls: [string, RsVal[]][] = [];
  if (arg.description) calls.push(['help', [rsStr(arg.description)]]);
  if (local && arg.acceptedValues?.length) calls.push(['value_parser', [valueParserList(arg.acceptedValues)]]);
  if (local && arg.arity) {
    const range = arityRange(arg.arity);
    if (range) calls.push(['num_args', [lit(range)]]);
  }
  if (arg.required === true) calls.push(['required', [lit('true')]]);
  return chain(`Arg::new(${JSON.stringify(arg.name)})`, calls);
}

File: crates/xyd_opencli2rust/src/flags.rs
//! clap Arg emission for flags + positionals — port of flags.ts.

use serde_json::Value;

use crate::model::{accepted_values_of, arity_of, ArityModel, FlagModel, FlagType};
use crate::rslit::{chain, json_str, lit, rs_str, RsVal};

/// A clap `.num_args(...)` range from an OpenCLI arity, or None when unconstrained.
fn arity_range(arity: &ArityModel) -> Option<String> {
    match (arity.minimum, arity.maximum) {
        (Some(min), Some(max)) => Some(format!("{min}..={max}")),
        (Some(min), None) => Some(format!("{min}..")),
        (None, Some(max)) => Some(format!("..={max}")),
        (None, None) => None,
    }
}

/// A clap `.value_parser([...])` argument literal from a list of accepted values.
fn value_parser_list(values: &[String]) -> RsVal {
    let list = values
        .iter()
        .map(|v| json_str(v))
        .collect::<Vec<_>>()
        .join(", ");
    lit(format!("[{list}]"))
}

pub fn render_flag_arg(f: &FlagModel) -> RsVal {
    let mut calls: Vec<(String, Vec<RsVal>)> = vec![("long".into(), vec![rs_str(&f.flag_name)])];

    let shorts: Vec<&String> = f
        .aliases
        .iter()
        .filter(|a| a.chars().count() == 1)
        .collect();
    let longs: Vec<&String> = f.aliases.iter().filter(|a| a.chars().count() > 1).collect();
    if !shorts.is_empty() {
        calls.push(("short".into(), vec![lit(format!("'{}'", shorts[0]))]));
    }
    if longs.len() == 1 {
        calls.push(("visible_alias".into(), vec![rs_str(longs[0])]));
    } else if longs.len() > 1 {
        let list = longs
            .iter()
            .map(|l| json_str(l))
            .collect::<Vec<_>>()
            .join(", ");
        calls.push(("visible_aliases".into(), vec![lit(format!("[{list}]"))]));
    }

    match f.flag_type {
        FlagType::Int => {
            calls.push(("value_parser".into(), vec![lit("clap::value_parser!(i64)")]));
        }
        FlagType::Float => {
            calls.push(("value_parser".into(), vec![lit("clap::value_parser!(f64)")]));
        }
        FlagType::Bool => {
            calls.push(("num_args".into(), vec![lit("0..=1")]));
            calls.push(("require_equals".into(), vec![lit("true")]));
            calls.push(("default_missing_value".into(), vec![rs_str("true")]));
            calls.push((
                "value_parser".into(),
                vec![lit("clap::value_parser!(bool)")],
            ));
        }
        FlagType::Slice => {
            calls.push(("action".into(), vec![lit("clap::ArgAction::Append")]));
            calls.push(("value_delimiter".into(), vec![lit("','")]));
        }
        // string | json | file: clap's default String parsing
        _ => {}
    }

    // Non-API leaf extras — additive and field-gated. `build_leaf_model` never sets
    // these, so x-openapi flags render exactly as before.
    if let Some(av) = f.accepted_values.as_ref().filter(|v| !v.is_empty()) {
        calls.push(("value_parser".into(), vec![value_parser_list(av)]));
    }
    if let Some(range) = f.arity.as_ref().and_then(arity_range) {
        calls.push(("num_args".into(), vec![lit(range)]));
    }

    if let Some(desc) = &f.description {
        calls.push(("help".into(), vec![rs_str(desc)]));
    }
    if f.required {
        calls.push(("required".into(), vec![lit("true")]));
    }
    if f.hidden {
        calls.push(("hide".into(), vec![lit("true")]));
    }

    chain(format!("Arg::new({})", json_str(&f.flag_name)), calls)
}

pub fn render_flag_args(flags: &[FlagModel]) -> Vec<RsVal> {
    flags.iter().map(render_flag_arg).collect()
}

/// A positional argument. `local` gates the non-API extras (`acceptedValues →
/// .value_parser`, `arity → .num_args`); the x-openapi call site passes `false`,
/// keeping API positionals byte-identical.
pub fn render_positional_arg(arg: &Value, local: bool) -> RsVal {
    let name = arg.get("name").and_then(|n| n.as_str()).unwrap_or("");
    let mut calls: Vec<(String, Vec<RsVal>)> = Vec::new();
    if let Some(desc) = arg.get("description").and_then(|d| d.as_str()) {
        calls.push(("help".into(), vec![rs_str(desc)]));
    }
    if local {
        if let Some(av) = accepted_values_of(arg) {
            calls.push(("value_parser".into(), vec![value_parser_list(&av)]));
        }
        if let Some(range) = arity_of(arg).as_ref().and_then(arity_range) {
            calls.push(("num_args".into(), vec![lit(range)]));
        }
    }
    if arg.get("required").and_then(|r| r.as_bool()) == Some(true) {
        calls.push(("required".into(), vec![lit("true")]));
    }
    chain(format!("Arg::new({})", json_str(name)), calls)
}
